### PR TITLE
Move PVF timeouts to executor environment parameters

### LINKED ISSUE #6793: Include PVF execution and preparation timeouts into executor environment parameters
PVF timeouts are crucial for determining candidate validity, and changing a timeout throughout the network may retrospectively affect the validity of already included candidates. Thus, such changes must be effective per session, not per network. Execution environment parameters are a natural place to keep track of timeout changes for the dispute period.

### PR DESCRIPTION
**Pull request by s0me0ne-unkn0wn**
Closes #6793 

PVF execution and preparation timeouts are stored in the executor environment parameters and are applied per session.

### CODE REVIEW

File: node/core/candidate-validation/src/lib.rs [hunk 1]
```diff
@@ -68,6 +69,34 @@ const PVF_EXECUTION_RETRY_DELAY: Duration = Duration::from_secs(3);
 #[cfg(test)]
 const PVF_EXECUTION_RETRY_DELAY: Duration = Duration::from_millis(200);
 
+/// Default PVF timeouts. Must never be changed! Use executor environment parameters in
+/// `session_info` pallet to adjust them
+
+/// For prechecking requests, the time period after which the preparation worker is considered
+/// unresponsive and will be killed.
+// NOTE: If you change this make sure to fix the buckets of `pvf_preparation_time` metric.
```
Comment (mrcnski) [inline node/core/candidate-validation/src/lib.rs:77]: These comments should probably be removed/updated in light of the "Must never be changed!" warning above. 😄

Comment (s0me0ne-unkn0wn) [reply node/core/candidate-validation/src/lib.rs:77]: You're right, I copied all the sensible comments to `enum PvfTimeoutType` and forgot to remove them here. Thank you!

File: node/core/pvf/src/prepare/queue.rs [hunk 2]
```diff
@@ -225,19 +220,18 @@ async fn handle_to_queue(queue: &mut Queue, to_queue: ToQueue) -> Result<(), Fat
...
-	preparation_timeout: Duration,
+	pvf: PvfExhaustive,
 ) -> Result<(), Fatal> {
 	gum::debug!(
 		target: LOG_TARGET,
-		validation_code_hash = ?pvf_with_params.code_hash(),
+		validation_code_hash = ?pvf.code_hash(),
 		?priority,
-		?preparation_timeout,
+		preparation_timeout = ?pvf.timeout(),
```
Comment (mrcnski) [inline node/core/pvf/src/prepare/queue.rs:229]: nit: Possible to log the whole `pvf` struct?

File: node/core/pvf/src/prepare/queue.rs [hunk 3]
```diff
@@ -503,18 +489,16 @@ pub fn start(
...
 	use assert_matches::assert_matches;
 	use futures::{future::BoxFuture, FutureExt};
 	use slotmap::SlotMap;
 	use std::task::Poll;
 
 	/// Creates a new PVF which artifact id can be uniquely identified by the given number.
-	fn pvf_with_params(descriminator: u32) -> PvfWithExecutorParams {
-		PvfWithExecutorParams::from_discriminator(descriminator)
+	fn pvf(descriminator: u32) -> PvfExhaustive {
+		PvfExhaustive::from_discriminator(descriminator)
```
Comment (mrcnski) [inline node/core/pvf/src/prepare/queue.rs:501]:
```suggestion
	fn pvf(discriminator: u32) -> PvfExhaustive {
		PvfExhaustive::from_discriminator(discriminator)
```

File: node/core/pvf/src/prepare/queue.rs [hunk 4]
```diff
@@ -636,33 +616,20 @@ mod tests {
...
-			preparation_timeout,
-		});
+		test.send_queue(ToQueue::Enqueue { priority, pvf: PvfExhaustive::from_discriminator(1) });
+		test.send_queue(ToQueue::Enqueue { priority, pvf: PvfExhaustive::from_discriminator(2) });
 		// Start a non-precheck preparation for this one.
 		test.send_queue(ToQueue::Enqueue {
 			priority,
-			pvf_with_params: PvfWithExecutorParams::from_discriminator(3),
-			preparation_timeout: LENIENT_PREPARATION_TIMEOUT,
+			pvf: PvfExhaustive::from_discriminator_and_timeout(3, TEST_PREPARATION_TIMEOUT * 3),
```
Comment (mrcnski) [inline node/core/pvf/src/prepare/queue.rs:632]: Maybe use `TEST_LENIENT_PREPARATION_TIMEOUT` instead? (Below too.)

File: node/core/pvf/src/execute/queue.rs [hunk 5]
```diff
@@ -45,8 +44,7 @@ use std::{
 /// re-spawn a new worker to execute the job immediately.
 /// To make any sense and not to break things, the value should be greater than minimal execution
 /// timeout in use, and less than the block time.
-const MAX_KEEP_WAITING: Duration =
-	Duration::from_millis(BACKING_EXECUTION_TIMEOUT.as_millis() as u64 * 2);
+const MAX_KEEP_WAITING: Duration = Duration::from_secs(4);
```
Comment (mrcnski) [inline node/core/pvf/src/execute/queue.rs:47]: Shouldn't we still base this off the backing timeout somehow?

File: node/core/pvf/src/host.rs [hunk 6]
```diff
@@ -507,9 +486,8 @@ async fn handle_execute_pvf(
 	awaiting_prepare: &mut AwaitingPrepare,
 	inputs: ExecutePvfInputs,
 ) -> Result<(), Fatal> {
-	let ExecutePvfInputs { pvf_with_params, execution_timeout, params, priority, result_tx } =
-		inputs;
-	let artifact_id = pvf_with_params.as_artifact_id();
+	let ExecutePvfInputs { pvf, execution_timeout, params, priority, result_tx } = inputs;
```
Comment (mrcnski) [inline node/core/pvf/src/host.rs:489]: Should the `execution_timeout` be part of the `pvf` now?

File: primitives/src/v2/mod.rs [hunk 7]
```diff
@@ -1689,6 +1689,35 @@ impl PvfCheckStatement {
...
+#[derive(Encode, Decode, TypeInfo, Clone, Copy, Debug, PartialEq, Eq)]
+pub enum PvfTimeoutType {
+	/// For prechecking requests, the time period after which the preparation worker is considered
+	/// unresponsive and will be killed.
+	PrecheckPreparation,
+
+	/// For execution and heads-up requests, the time period after which the preparation worker is
+	/// considered unresponsive and will be killed. More lenient than the timeout for prechecking
+	/// to prevent honest validators from timing out on valid PVFs.
+	LenientPreparation,
```
Comment (mrcnski) [inline primitives/src/v2/mod.rs:1702]: Looks like this variant is not used anywhere, AFAICT.

**Review by mrcnski — CHANGES_REQUESTED** [review]
Really like this change! I mostly left a bunch of nits and a few questions.

File: primitives/src/v2/mod.rs (hunk 7 above)
Comment (s0me0ne-unkn0wn) [reply primitives/src/v2/mod.rs:1702]: Good catch, sounds like a real problem, that means I'm using a wrong timeout somewhere. Will investigate.

File: node/core/pvf/src/host.rs (hunk 6 above)
Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/host.rs:489]: I believe it shouldn't as the execution timeouts are per job, not per artifact.

Comment (mrcnski) [reply node/core/pvf/src/host.rs:489]: Looks like in `validate_candidate_with_retry` we construct the `pvf` from the timeout
:
```rs
let pvf = PvfExhaustive::from_code(raw_validation_code, executor_params, timeout);
```

but then we still pass along the PVF and timeout:

```rs
validation_result = self.validate_candidate(pvf, timeout, params.encode()).await;
```

File: node/core/pvf/src/pvf.rs [hunk 8]
```diff
@@ -23,26 +23,31 @@ use std::{
...
+/// A struct that carries code of a parachain validation function, its hash, a corresponding
+/// set of executor parameters, and a preparation timeout.
 ///
 /// Should be cheap to clone.
 #[derive(Clone, Encode, Decode)]
-pub struct PvfWithExecutorParams {
+pub struct PvfExhaustive {
 	pub(crate) code: Arc<Vec<u8>>,
 	pub(crate) code_hash: ValidationCodeHash,
 	pub(crate) executor_params: Arc<ExecutorParams>,
```
[the remark this answers is not in the document]
Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/pvf.rs:40]: It is needed, as the Wasm stack limit is enforced during the preparation stage through instrumentation (see `prepare_blob_for_compilation()` in Substrate executor code). Also, we'll keep a preparation memory limit inside `ExecutorParams` in the future.

File: primitives/src/v2/mod.rs [hunk 9]
```diff
@@ -1689,6 +1689,35 @@ impl PvfCheckStatement {
...
+	/// absence of extremely large disparities between hardware, blocks that pass backing are
+	/// considered executable by approval checkers or dispute participants.
+	ApprovalExecution,
+}
+
+impl Default for PvfTimeoutType {
+	fn default() -> Self {
+		Self::BackingExecution
+	}
+}
```
[the remark this answers is not in the document]
Comment (s0me0ne-unkn0wn) [reply primitives/src/v2/mod.rs:1719]: Nice catch, `default()` is only supposed to be used in tests, so it should be under `#[cfg(test)]`

File: node/core/pvf/src/host.rs (hunk 6 above)
Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/host.rs:489]: Exactly. The timeout inside `PvfExhaustive` is the preparation timeout, and the one passed to `validate_candidate()` is the execution timeout. PVF host may or may not have an already prepared artifact by the time of validation request, so it needs both. But the naming is dubious, I think they should be renamed.

File: node/core/pvf/src/pvf.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (mrcnski) [reply node/core/pvf/src/pvf.rs:40]: OK, thanks!

File: node/core/pvf/src/host.rs (hunk 6 above)
Comment (mrcnski) [reply node/core/pvf/src/host.rs:489]: Oh, I see. Yeah, renaming it to `preparation_timeout` would be better. PVFs can be either prepared or executed so we should distinguish what the timeout is for in its name.

File: node/core/pvf/src/pvf.rs [hunk 10]
```diff
@@ -23,26 +23,31 @@ use std::{
...
+#[cfg(test)]
+use crate::host::tests::TEST_PREPARATION_TIMEOUT;
+
+/// A struct that carries code of a parachain validation function, its hash, a corresponding
+/// set of executor parameters, and a preparation timeout.
 ///
 /// Should be cheap to clone.
 #[derive(Clone, Encode, Decode)]
-pub struct PvfWithExecutorParams {
+pub struct PvfExhaustive {
```
[the remark this answers is not in the document]
Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/pvf.rs:37]: Here's a problem. There is already something called "PVF params", those are params passed to the PVF itself during the execution. So we called environment params the "executor params" and now, if we called the whole thing `PvfWithParams` it would be twice dubious. So if you have any better naming ideas for "PVF-code-blob-and-everything-needed-to-prepare-an-artifact", you're welcome, I'm not a big fan of that "Exhaustive" too, I just failed to invent something more meaningful, and I didn't want to fall back to just `Pvf`, because it is not just PVF anymore.

File: node/core/pvf/src/prepare/queue.rs (hunk 2 above)
Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/prepare/queue.rs:229]: With the code blob? :scream: 
IIUC, we need the validation code hash to be named exactly `validation_code_hash` in logs to be observable. The executor parameters are logged on `debug` level in the candidate validation subsys when validation is requested. So the only meaningful field here is `timeout`, I believe.

File: node/core/pvf/src/pvf.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (mrcnski) [reply node/core/pvf/src/pvf.rs:37]: Yes, it is an awkward situation. Naming is a hard problem. Maybe... `PvfWithPreparationData`? Haha, not great, but better I think. Or maybe `PvfPreparationData` or even `PreparationData`. If I'm reading code and I see `PvfExhaustive` by itself, it doesn't tell me anything about what the struct does.

File: node/core/pvf/src/prepare/queue.rs (hunk 2 above)
Comment (mrcnski) [reply node/core/pvf/src/prepare/queue.rs:229]: Ah gotcha. I just thought it could be more future-proof in case more fields were added. Maybe it could impl `Display`, which would only print a very truncated code blob? Just an idea, feel free to resolve if it's a bad one.

File: primitives/src/v2/mod.rs (hunk 7 above)
Comment (s0me0ne-unkn0wn) [reply primitives/src/v2/mod.rs:1702]: Found it, I've mixed preparation and execution timeouts in several places which is really bad and makes me think that having separate `PvfPreparationTimeoutKind` and `PvfExecutionTimeoutKind` would be a good idea that could help us to avoid such errors in the future. The only drawback is that really long naming (as usual).

Comment (s0me0ne-unkn0wn) [reply primitives/src/v2/mod.rs:1702]: If we agree to abbreviate "preparation" as "prep" and "execution" as "exec" everywhere, it would be less painful

Comment (mrcnski) [reply primitives/src/v2/mod.rs:1702]: I am always in favor of using more types to ensure correctness in the code. ⭐️ 

I don't think those names are *that* long, pretty sure we have longer names than that, but I wouldn't mind shortening to "prep" and "exec" everywhere -- should be pretty evident what those mean.

File: node/service/chain-specs/westend.json [hunk 11]
```diff
@@ -18,8 +18,8 @@
     "/dns/boot-node.helikon.io/tcp/7082/wss/p2p/12D3KooWRFDPyT8vA8mLzh6dJoyujn4QNjeqi6Ch79eSMz9beKXC",
     "/dns/westend.bootnode.amforc.com/tcp/30333/p2p/12D3KooWJ5y9ZgVepBQNW4aabrxgmnrApdVnscqgKWiUu4BNJbC8",
     "/dns/westend.bootnode.amforc.com/tcp/30334/wss/p2p/12D3KooWJ5y9ZgVepBQNW4aabrxgmnrApdVnscqgKWiUu4BNJbC8",
-    "/dns/westend-bootnode.polkadotters.com/tcp/30333/p2p/12D3KooWQiGqJxKmaRboMN2ATkW22R6SiXrc6Tp1APGwequhzuJU",
-    "/dns/westend-bootnode.polkadotters.com/tcp/30334/wss/p2p/12D3KooWQiGqJxKmaRboMN2ATkW22R6SiXrc6Tp1APGwequhzuJU",
+    "/dns/westend-bootnode.polkadotters.com/tcp/30333/p2p/12D3KooWHPHb64jXMtSRJDrYFATWeLnvChL8NtWVttY67DCH1eC5",
+    "/dns/westend-bootnode.polkadotters.com/tcp/30334/wss/p2p/12D3KooWHPHb64jXMtSRJDrYFATWeLnvChL8NtWVttY67DCH1eC5",
```
Comment (ordian) [inline node/service/chain-specs/westend.json:22]: rebase went wrong? this looks unrelated

Comment (s0me0ne-unkn0wn) [reply node/service/chain-specs/westend.json:22]: Terribly wrong :/
Looking how to roll it back

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2492752

File: node/core/candidate-validation/src/lib.rs [hunk 12]
```diff
@@ -68,6 +69,13 @@ const PVF_EXECUTION_RETRY_DELAY: Duration = Duration::from_secs(3);
 #[cfg(test)]
 const PVF_EXECUTION_RETRY_DELAY: Duration = Duration::from_millis(200);
 
+/// Default PVF timeouts. Must never be changed! Use executor environment parameters in
+/// `session_info` pallet to adjust them. See also `PvfTimeoutKind` docs.
+const DEFAULT_PRECHECK_PREPARATION_TIMEOUT: Duration = Duration::from_secs(60);
```
Comment (mrcnski) [inline node/core/candidate-validation/src/lib.rs:74]: Small nit, this docstring would be attached to `DEFAULT_PRECHECK_PREPARATION_TIMEOUT` and not the following items. Change to:

```suggestion
// Default PVF timeouts. Must never be changed! Use executor environment parameters in
// `session_info` pallet to adjust them. See also `PvfTimeoutKind` docs.
const DEFAULT_PRECHECK_PREPARATION_TIMEOUT: Duration = Duration::from_secs(60);
```

**Review by mrcnski — APPROVED** [review]
Awesome!!

**Review by ordian — APPROVED** [review]
LVGTM

File: primitives/src/vstaging/executor_params.rs [hunk 13]
```diff
@@ -39,6 +39,10 @@ pub enum ExecutorParam {
 	/// Max. amount of memory the preparation worker is allowed to use during
 	/// pre-checking, in bytes
 	PrecheckingMaxMemory(u64),
+	/// PVF preparation timeouts, millisec
+	PvfPrepTimeout(PvfPrepTimeoutKind, u64),
```
Comment (ordian) [inline primitives/src/vstaging/executor_params.rs:43]: Does it make sense to annotate with `#[codec(index = X)]` to be explicit about backwards compatibility?

File: primitives/src/vstaging/executor_params.rs [hunk 14]
```diff
@@ -92,6 +112,16 @@ impl ExecutorParams {
 	pub fn hash(&self) -> ExecutorParamsHash {
 		ExecutorParamsHash(BlakeTwo256::hash(&self.encode()))
 	}
+
+	timeout_by_kind! {
```
Comment (eskimor) [inline primitives/src/vstaging/executor_params.rs:116]: Not a fan of macro usage here: Obfuscates the interface, would definitely prefer a bit of code duplication here over a macro.

**Review by eskimor — APPROVED** [review]
Nice, great work @s0me0ne-unkn0wn !

**Comment by s0me0ne-unkn0wn** [comment]
bot merge
